naptime-graphql/src/main/scala/org/coursera/naptime/ari/graphql/schema/FieldBuilder.scala

```diff
@@ -0,0 +1,167 @@
+import com.linkedin.data.schema.IntegerDataSchema
+import com.linkedin.data.schema.LongDataSchema
+import com.linkedin.data.schema.MapDataSchema
+import com.linkedin.data.schema.RecordDataSchema
+import com.linkedin.data.schema.RecordDataSchema.{Field => RecordDataSchemaField}
+import com.linkedin.data.schema.StringDataSchema
+import com.linkedin.data.schema.TyperefDataSchema
+import com.linkedin.data.schema.UnionDataSchema
+import org.coursera.naptime.ari.graphql.SangriaGraphQlContext
+import org.coursera.naptime.ari.graphql.types.NaptimeTypes
+import sangria.schema.BooleanType
+import sangria.schema.Context
+import sangria.schema.Field
+import sangria.schema.FloatType
+import sangria.schema.IntType
+import sangria.schema.ListType
+import sangria.schema.LongType
+import sangria.schema.OptionType
+import sangria.schema.StringType
+import sangria.schema.Value
+
+import scala.collection.JavaConverters._
+
+
+object FieldBuilder {
+
+  type ResolverType = Context[SangriaGraphQlContext, DataMap] => Value[SangriaGraphQlContext, Any]
+
+  def buildField(
+      schemaMetadata: SchemaMetadata,
+      field: RecordDataSchemaField,
+      namespace: Option[String],
+      fieldNameOverride: Option[String] = None): Field[SangriaGraphQlContext, DataMap] = {
+    type ResolverType = Context[SangriaGraphQlContext, DataMap] => Value[SangriaGraphQlContext, Any]
+
+    val relatedResourceOption = field.getProperties.asScala.get("related").map(_.toString)
+
+    val fieldName = fieldNameOverride.getOrElse(field.getName)
+
+    val fieldDoc = if (field.getDoc.isEmpty) None else Some(field.getDoc)
+
+    val sangriaField = (relatedResourceOption, field.getType) match {
+
+      // Related resource in a list
+      case (Some(relatedResourceName), _: ArrayDataSchema) =>
+        NaptimePaginatedResourceField.build(schemaMetadata, relatedResourceName, fieldName)
+
+      // Single related resource
+      case (Some(relatedResourceName), _) =>
+        NaptimeResourceField.build(schemaMetadata, relatedResourceName, fieldName)
+
+      // Passthrough Exempt types (fallback to DataMap)
+      case (None, recordDataSchema: RecordDataSchema)
+        if recordDataSchema.getProperties.asScala.get("passthroughExempt").contains(true) =>
+
+        Field.apply[SangriaGraphQlContext, DataMap, Any, Any](
+          name = fieldName,
+          fieldType = NaptimeTypes.DataMapType,
+          resolve = context => context.value.getDataMap(fieldName))
+
+      // Complex types
+      case (None, recordDataSchema: RecordDataSchema) =>
+        NaptimeRecordField.build(schemaMetadata, recordDataSchema, fieldName, namespace)
+
+      case (None, enumDataSchema: EnumDataSchema) =>
+        NaptimeEnumField.build(enumDataSchema, fieldName)
+
+      case (None, unionDataSchema: UnionDataSchema) =>
+        NaptimeUnionField.build(schemaMetadata, unionDataSchema, fieldName, namespace)
+
+      case (None, typerefDataSchema: TyperefDataSchema) =>
```

Comment:
For typeref data schema, I think you should call `getDereferencedDataSchema` instead of `getRef`. Have you been able to test this code path? (esp. with typeref's of typerefs.)